### PR TITLE
Split type-unsafe InternalQueryCache into 4 statically typed caches

### PR DESCRIPTION
**Pull request by SergeiPavlov**
Also eliminated boxing for keys and values

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/EntitySetBase.cs [hunk 1]
```diff
@@ -923,17 +922,15 @@ private static Key GetOwnerKey(Persistent owner)
...
     {
       EnsureOwnerIsNotRemoved();
-      object key = new Pair<object, FieldInfo>(entitySetCachingRegion, Field);
-      Func<object, object> generator = k => BuildEntitySetTypeState(k, this);
-      return (EntitySetTypeState) Session.StorageNode.InternalQueryCache.GetOrAdd(key, generator);
+      return Session.StorageNode.InternalEntitySetCache.GetOrAdd(Field, BuildEntitySetTypeState);
     }
 
-    private static EntitySetTypeState BuildEntitySetTypeState(object key, EntitySetBase entitySet)
+    private EntitySetTypeState BuildEntitySetTypeState(object key)
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/EntitySetBase.cs:928]: Btw, with help of the https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-5.0#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__1__0_System_Func__0___0__1____0_ overload we still can keep this method factory static or convert it to a static delegate as you did in other cases.

File: Orm/Xtensive.Orm/Orm/StorageNode.cs [hunk 2]
```diff
@@ -44,11 +47,25 @@ public sealed class StorageNode : ISessionSource
     /// </summary>
     public TypeIdRegistry TypeIdRegistry { get; private set; }
 
-    internal ConcurrentDictionary<object, object> InternalQueryCache { get; private set; }
+    internal ConcurrentDictionary<(TypeInfo, LockMode, LockBehavior), ExecutableProvider> InternalExecutableProviderCache { get; } =
+      new ConcurrentDictionary<(TypeInfo, LockMode, LockBehavior), ExecutableProvider>();
 
-    internal ConcurrentDictionary<SequenceInfo, object> KeySequencesCache { get; private set; }
+    internal ConcurrentDictionary<EntityGroupTask.CacheKey, CompilableProvider> InternalRecordSetCache { get; } =
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/StorageNode.cs:53]: Maybe it makes sense to extract the nested `CacheKey` structure and rename it to e.g. `RecordSetCacheKey` to increase readability.

File: Orm/Xtensive.Orm/Orm/StorageNode.cs [hunk 3]
```diff
@@ -44,11 +47,25 @@ public sealed class StorageNode : ISessionSource
...
+      new ConcurrentDictionary<EntityGroupTask.CacheKey, CompilableProvider>();
 
-    internal ConcurrentDictionary<PersistRequestBuilderTask, ICollection<PersistRequest>> PersistRequestCache { get; private set; }
+    internal ConcurrentDictionary<EntitySetTask.CacheKey, CompilableProvider> InternalItemsQueryCache { get; } =
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/StorageNode.cs:56]: The same here.

**Review by AlexUstinov — APPROVED** [review]

File: Orm/Xtensive.Orm/Orm/EntitySetBase.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/EntitySetBase.cs]: Yes, you are right
done

File: Orm/Xtensive.Orm/Orm/StorageNode.cs (hunk 3 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/StorageNode.cs]: done

File: Orm/Xtensive.Orm/Orm/StorageNode.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/StorageNode.cs]: done

File: Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntitySetTask.cs [hunk 4]
```diff
@@ -19,47 +19,43 @@
 
 namespace Xtensive.Orm.Internals.Prefetch
 {
-  [Serializable]
-  internal sealed class EntitySetTask : IEquatable<EntitySetTask>
+  internal struct ItemsQueryCacheKey : IEquatable<ItemsQueryCacheKey>
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntitySetTask.cs:22]: readonly struct?

File: Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntityGroupTask.cs [hunk 5]
```diff
@@ -15,59 +15,55 @@
 
 namespace Xtensive.Orm.Internals.Prefetch
 {
-  [Serializable]
-  internal sealed class EntityGroupTask : IEquatable<EntityGroupTask>
+  internal struct RecordSetCacheKey : IEquatable<RecordSetCacheKey>
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntityGroupTask.cs:18]: readonly struct?

File: Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntitySetTask.cs (hunk 4 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntitySetTask.cs]: done

File: Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntityGroupTask.cs (hunk 5 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Internals/Prefetch/EntityGroupTask.cs]: done

**Pull request merged**